codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java

```diff
@@ -233,35 +232,32 @@ void onUnknownFrame0(ChannelHandlerContext ctx, byte frameType, int streamId, Ht
         @Override
         public int onDataRead(final ChannelHandlerContext ctx, int streamId, ByteBuf data,
                 int padding, boolean endOfStream) throws Http2Exception {
-            // Check if we received a data frame for a stream which is half-closed
             Http2Stream stream = connection.requireStream(streamId);
```

Comment:
@buchgr - I agree the connection error may not be the right thing to do here. [section 6.1](https://tools.ietf.org/html/draft-ietf-httpbis-http2-17#section-6.1):

> If a DATA frame is received whose
   stream is not in "open" or "half closed (local)" state, the recipient
   MUST respond with a stream error (Section 5.4.2) of type
   STREAM_CLOSED.

We do that [DefaultHttp2ConnectionDecoder.java#L255](https://github.com/buchgr/netty/blob/fix_goaway/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java#L255) but we have to get the stream first to get here...

Some things we should watch out for is when is it OK to send a RST_STREAM frame.

[section 6.4](https://tools.ietf.org/html/draft-ietf-httpbis-http2-17#section-6.4):

> RST_STREAM frames MUST NOT be sent for a stream in the "idle" state.

Also we may want to be careful about sending on streams that *may* have existed...but in this case I think we are OK sending the RST_STREAM [section 5.4.2](https://tools.ietf.org/html/draft-ietf-httpbis-http2-17#section-5.4.2):

> Normally, an endpoint SHOULD NOT send more than one RST_STREAM frame
   for any stream.  However, an endpoint MAY send additional RST_STREAM
   frames if it receives frames on a closed stream after more than a
   round-trip time.  This behavior is permitted to deal with misbehaving
   implementations.